### ISSUE TITLE
ER: Devs running WSL Ubuntu cannot run `docker-compose up`

### ISSUE THREAD

**Issue by roslynwythe** [hackforla/website#5647]
See comment below https://github.com/hackforla/website/issues/5647#issuecomment-1972279146 for instructions for how to resolve ER

### Emergent Requirement - Problem
Several new devs running WSL Ubuntu have been blocked by permission errors when running `docker-compose up`

### Issue you discovered this emergent requirement in
see Slack thread copied into https://github.com/hackforla/website/issues/5647#issuecomment-1797033119

### Did you have to do something temporarily
- [x] YES - we have to keep fielding slack messages
- [ ] NO

### Who was involved
@freaky4wrld @jaasonw @klei0229  Matt Pereira

### What happens if this is not addressed
Potential contributors may become frustrated and quit Hack for LA

### Resources
see comment by @klei0229 https://github.com/hackforla/website/issues/5219#issuecomment-1691423034
see https://github.com/hackforla/website/issues/5647#issuecomment-1797033119
also see other comments in this issue

### Recommended Action Items
- [x] Make a new issue
- [ ] Discuss with team
- [ ] Let a Team Lead know

### Potential solutions [draft]
Create an issue to add to the contributing.md file with instructions for how to find your version of Go. 

Fang says here https://github.com/hackforla/website/issues/5647#issuecomment-1912793042 Anything less than 1.20 has a problem.

**Comment by freaky4wrld** [comment]
Maybe the situation might differ as the above mentioned solution is not perfect @klei0229 was able to resolve using [this](https://github.com/hackforla/website/issues/5219#issuecomment-1691423034) approach, and for me it's different like the issue that I have seen is my ` host machine UID: 1000 ` but for my ` Docker Container it's 0 `. I'm still stuck with this issue and trying to find out the solution to run the site locally.

My `Docker Container's UID`

![Screenshot from 2023-10-03 14-24-44](https://github.com/hackforla/website/assets/134206855/57c146ef-7566-4beb-8d0c-b2208acc7b7a)

My `host machine UID`

![Screenshot from 2023-10-03 14-27-04](https://github.com/hackforla/website/assets/134206855/8689e9e5-842d-4df6-8c93-f2b8f4d40b09)

**Comment by ExperimentsInHonesty** [comment]
@roslynwythe please copy in the conversations from Slack, so we don't loose the history.

**Comment by roslynwythe** [comment]
The original Slack thread:

Freaky Wrld
I've been trying to complete my prework checklist and got stuck while hosting the site locally through docker can anybody point out to me where to look for solution...... refer the below screenshot, my current working directory is website, I'm trying to execute the docker-compose up command and the Docker Desktop is running...... what I'm doing wrong [here](https://stackoverflow.com/questions/57503011/unable-to-build-cloned-jekyll-site-jekyll-3-8-5-error-permission-denied-d), I looked up here for the solution, but it turned out my current UID is already 1000, but still I tried to provide the environment variables in the docker-compose.yml file, and found that's not working either.
4 files
 
Screenshot from 2023-09-29 13-50-03.png


![Screenshot from 2023-09-29 13-50-03](https://github.com/hackforla/website/assets/5314153/bcd61b0f-719b-4e7c-9a36-625b40456e9f)


Screenshot from 2023-09-29 14-02-32.png

![Screenshot from 2023-09-29 14-02-32](https://github.com/hackforla/website/assets/5314153/0beb1ebe-4140-4962-94a6-e0d0c5923988)

Screenshot from 2023-09-29 14-05-44.png

![Screenshot from 2023-09-29 14-05-44](https://github.com/hackforla/website/assets/5314153/4fea2054-3c54-4a38-a272-5c6306ece446)

Screenshot from 2023-09-29 14-06-27.png
![Screenshot from 2023-09-29 14-06-27](https://github.com/hackforla/website/assets/5314153/fadab381-2d88-44b3-bcac-ccae3c52df30)

Stack OverflowStack Overflow
[Unable to build cloned Jekyll site - jekyll 3.8.5 | Error: Permission denied @ dir_s_mkdir - /srv/jekyll/_site](https://stackoverflow.com/questions/57503011/unable-to-build-cloned-jekyll-site-jekyll-3-8-5-error-permission-denied-d)



Freaky Wrld
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1695977083880989?thread_ts=1695976609.442539&cid=C4UM52W93)
Btw I'm on Ubuntu 22.04.3


Fang
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696018726252159?thread_ts=1695976609.442539&cid=C4UM52W93)
Do you have docker installed as rootless?


Fang
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696018954663849?thread_ts=1695976609.442539&cid=C4UM52W93)
The image sets up a non-root user to run the server and docker rootless config doesn’t support that.


Fang
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696019073964849?thread_ts=1695976609.442539&cid=C4UM52W93)
Either use podman for rootless or run docker server as root.


Fang
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696019109764429?thread_ts=1695976609.442539&cid=C4UM52W93)
I’m running my docker daemon rootless so I’m aware of the issue. (edited) 


Freaky Wrld
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696032932517429?thread_ts=1695976609.442539&cid=C4UM52W93)
I don't have any knowledge about docker, installed it following the documentation


Fang
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696033227451369?thread_ts=1695976609.442539&cid=C4UM52W93)
Oh, then it’s probably a different issue.


Roslyn Wythe
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696271549705979?thread_ts=1695976609.442539&cid=C4UM52W93)
[@Fang](https://hackforla.slack.com/team/UCE01B4AX)
 I'm curious why do you run docker daemon rootless?


Fang
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696413363153069?thread_ts=1695976609.442539&cid=C4UM52W93)
[@Roslyn Wythe](https://hackforla.slack.com/team/U046PD8UT55)
 Seems like I'm the only one here that does this so maybe it's not that important to do or it's not yet standard practice. I think it's pretty safe for the website team's ghpages docker image to run on a docker daemon running as root. But here's the motivation for running it rootless, but actual usage can be complicated depending on the image being used.
It’s supposedly safer for running docker containers, because it's possible for a compromised container to compromise the host machine running docker as root. Docker has protections against that (seccomp filters out admin-level syscalls, AppArmor does something too) but running the whole thing as a limited user is a more certain guarantee. [Here's the explanation](https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-12-run-docker-in-root-less-mode) for it from an organization dedicated to web security. Here's [docker's docs on security](https://docs.docker.com/engine/security/) with a page on rootless docker.
But here's the problem with rootless: Docker's rootless mode only supports running as root inside the container because that's what it assumes. It has issues with file permissions when running containers as a non-root user (rootless image) like the ghpage image the website team uses does. The rootless image is a safety measure to limit the chances of running something in the container that will compromise the host running a rootful docker daemon. So I think it's still pretty safe to run the ghpages image with a rootful docker daemon.
It's confusing because the rootless daemon doesn't support rootless images, so you have to choose one or the other. [Some software](https://help.passbolt.com/faq/hosting/how-to-use-rootless-images) provide docker images that run as root and non-root, which is a way to satisfy both camps.
It's enough hassle to run security-minded images that at some point I'm planning to switch back to the rootful docker daemon if podman doesn't work well. Podman is supposed to be a drop-in replacement for docker to run docker containers, only it's daemon-less and has better user-mapping support than docker. It should work according to [this comment](https://github.com/microsoft/vscode-remote-release/issues/4646#issuecomment-1320313802).

**Comment by roslynwythe** [comment]
> @roslynwythe please copy in the conversations from Slack, so we don't loose the history.

@ExperimentsInHonesty - Done.  I just learned a few days ago that we cannot access old Slack messages.  Thanks!

**Comment by ExperimentsInHonesty** [comment]
I approve this being made into an issue.  For whoever makes it into an issue, ~please add a prerequisite that the person working on the solution and~ the people reviewing the PR will have a machine running on Ubuntu

**Comment by roslynwythe** [comment]
@jaasonw @freaky4wrld @fyliu  I saw a [bug report for the docker.io-app package in Ubuntu](https://bugs.launchpad.net/ubuntu/+source/docker.io-app/+bug/2029523) that may describe the source of this problem.  If so, the problem could be resolved by either using the newest version of the package or going back to an older version: version 20.10.21-0ubuntu1~20.04.2

**Comment by fyliu** [comment]
@roslynwythe that's great work!

Feels like it's in the `docker` client, since it only happens to some people.

What a coincidence that people onboarded during the timeframe where `docker` was built against a buggy version of `go` that caused ownership information to be lost when unpacking the `docker` layers.

So it looks like the easiest way to know if a fix is necessary is to run `docker version | grep "Go"` to see what version of `go` is used. Anything less than `1.20` has a problem. Maybe new volunteers won't encounter the problem since they'd be installing the latest `docker`.

**Comment by roslynwythe** [comment]
Thank you @fyliu.  I'll try to locate other Hfla developers who are experiencing this problem, to see if upgrading the Ubuntu docker client resolves the problem.

**Comment by ExperimentsInHonesty** [comment]
This ER will be resolved when someone makes an issue to add to the contributing.md file with instructions for how to find your version of Go

Fang says here https://github.com/hackforla/website/issues/5647#issuecomment-1912793042 Anything less than 1.20 has a problem.

**Comment by HackforLABot** [comment]
Hi @k-cardon, thank you for taking up this issue! Hfla appreciates you :)

Do let fellow developers know about your:-
i. **Availability:** (When are you available to work on the issue/answer questions other programmers might have about your issue?)
ii. **ETA:** (When do you expect this issue to be completed?)

You're awesome!

P.S. - You may not take up another issue until this issue gets merged (or closed). Thanks again :)

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty @k-cardon is interested in this issue.  She has been waiting for a level 2 issue writing issue for good first issue but there are none of those so I thought this one would be a good choice for her.

**Comment by HackforLABot** [comment]
Hi @k-cardon, thank you for taking up this issue! Hfla appreciates you :)

Do let fellow developers know about your:-
i. **Availability:** (When are you available to work on the issue/answer questions other programmers might have about your issue?)
ii. **ETA:** (When do you expect this issue to be completed?)

You're awesome!

P.S. - You may not take up another issue until this issue gets merged (or closed). Thanks again :)

**Comment by k-cardon** [comment]
Availability: usually evenings / weekends / Tuesday afternoons

ETA: within one week

**Comment by k-cardon** [comment]
@roslynwythe just an update that @ExperimentsInHonesty gave permission for me to work on this one!

**Comment by k-cardon** [comment]
I made the issue from this ER and it is ready for review:
- #7642

**Comment by ExperimentsInHonesty** [comment]
@k-cardon Thank you for working on this.  Your issue has been approved
